controllers/imagerepository_controller.go

```diff
@@ -72,55 +73,85 @@ func (r *ImageRepositoryReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
 
 	ref, err := name.ParseReference(imageRepo.Spec.Image)
 	if err != nil {
-		imageRepo.Status.LastError = err.Error()
-		if err := r.Status().Update(ctx, &imageRepo); err != nil {
-			return ctrl.Result{}, err
+		status := imagev1alpha1.SetImageRepositoryReadiness(
+			imageRepo,
+			corev1.ConditionFalse,
+			imagev1alpha1.ImageURLInvalidReason,
+			err.Error(),
+		)
+		if err := r.Status().Update(ctx, &status); err != nil {
+			return ctrl.Result{Requeue: true}, err
 		}
 		log.Error(err, "Unable to parse image name", "imageName", imageRepo.Spec.Image)
 		return ctrl.Result{}, nil
 	}
 
-	canonicalName := ref.Context().String()
-	imageRepo.Status.CanonicalImageName = canonicalName
+	imageRepo.Status.CanonicalImageName = ref.Context().String()
 
 	now := time.Now()
-	ok, when := r.shouldScan(&imageRepo, now)
+	ok, when := r.shouldScan(imageRepo, now)
 	if ok {
 		ctx, cancel := context.WithTimeout(ctx, scanTimeout)
 		defer cancel()
-		tags, err := remote.ListWithContext(ctx, ref.Context()) // TODO: auth
-		if err != nil {
-			imageRepo.Status.LastError = err.Error()
-			if err = r.Status().Update(ctx, &imageRepo); err != nil {
-				return ctrl.Result{}, err
-			}
+
+		reconciledRepo, reconcileErr := r.scan(ctx, imageRepo, ref)
+		if err = r.Status().Update(ctx, &reconciledRepo); err != nil {
+			return ctrl.Result{Requeue: true}, err
 		}
 
-		imageRepo.Status.LastScanTime = &metav1.Time{Time: now}
-		imageRepo.Status.LastScanResult.TagCount = len(tags)
-		imageRepo.Status.LastError = ""
-		// share the information in the database
-		r.Database.SetTags(canonicalName, tags)
-		log.Info("successful scan", "tag count", len(tags))
-		if err = r.Status().Update(ctx, &imageRepo); err != nil {
-			return ctrl.Result{}, err
+		if reconcileErr != nil {
+			return ctrl.Result{Requeue: true}, reconcileErr
```

Comment:
This will still show up as a reconciliation error, when it might be because e.g., the registry scan failed (i.e., a failure to make progress). I see that e.g., the source-controller does a similar thing -- perhaps it's worth having a (GH) discussion about it?

Reply:
In my mind a reconciliation error should bubble up no matter what the controller does: registry scan, git cloning, kubectl apply, helm install, etc. If we don't consider those as reconciliation errors then we should delete the monitoring stack as it becomes useless.

Reply:
https://github.com/fluxcd/toolkit/discussions/164

> If we don't consider those as reconciliation errors then we should delete the monitoring stack as it becomes useless

That's a bit over-dramatic. There's nothing stopping you recording your own metrics for failures to converge.

Reply:
See https://github.com/fluxcd/image-reflector-controller/pull/24#issuecomment-677647182